primefaces/src/main/java/org/primefaces/model/LazyTreeNode.java

```diff
@@ -66,49 +71,58 @@ public int getChildCount() {
 
     @Override
     public boolean isLeaf() {
-        return !getData().isDirectory();
+        return leafFunction.apply(getData());
     }
 
-    private void lazyLoad() {
-        if (!lazyLoaded) {
-            lazyLoaded = true;
+    public boolean isLoaded() {
+        return loaded;
+    }
 
-            String parentId = ((FileInfo) getData()).getPath();
+    protected void lazyLoad() {
+        if (!loaded) {
+            loaded = true;
 
-            List<LazyLoadingTreeNode> childNodes = loadFunction.apply(parentId).stream()
-                    .map(f -> new LazyLoadingTreeNode(f, loadFunction)).collect(Collectors.toList());
+            List<T> childData = loadFunction.apply(getData());
+            List<LazyTreeNode<T>> childNodes = childData
+                    .stream()
+                    .map(f -> new LazyTreeNode<>(f, loadFunction, leafFunction))
+                    .collect(Collectors.toList());
             super.getChildren().addAll(childNodes);
         }
     }
 
     @Override
-    protected List<TreeNode<FileInfo>> initChildren() {
-        return new LazyLoadingTreeNodeChildren(this);
+    protected List<TreeNode<T>> initChildren() {
```

Comment:
I think we said that data could have a different type between parent and children?

Reply:
see DefaulTreeNode:

protected List<TreeNode<T>> initChildren() {

i didnt refeactor anything
maybe better placed in your refactoring PR